File: deck.py
"""
Deck/Shoe management for Blackjack simulation.
Supports both finite 8-deck shoe and infinite deck modes.
"""

import random
from typing import List, Optional
import numpy as np


# Card values: 2-10 face value, J/Q/K = 10, A = 11 (handled specially)
CARD_VALUES = [2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 11]  # 2-10, J, Q, K, A


class Shoe:
    """
    8-deck shoe with configurable penetration.
    Shuffles when penetration threshold is reached.
    """

    def __init__(self, num_decks: int = 8, penetration: float = 0.5):
        """
        Initialize shoe.

        Args:
            num_decks: Number of decks in shoe (default 8)
            penetration: Fraction of shoe dealt before reshuffle (default 0.5)
        """
        self.num_decks = num_decks
        self.penetration = penetration
        self.cards: List[int] = []
        self.cut_card_position: int = 0
        self.shuffle()

    def shuffle(self) -> None:
        """Shuffle all cards back into the shoe."""
        # Each deck has 4 of each card value (4 suits)
        single_deck = CARD_VALUES * 4
        self.cards = single_deck * self.num_decks
        random.shuffle(self.cards)
        # Cut card position - reshuffle after this many cards dealt
        self.cut_card_position = int(len(self.cards) * self.penetration)

    def needs_shuffle(self) -> bool:
        """Check if shoe needs reshuffling based on penetration."""
        cards_dealt = (self.num_decks * 52) - len(self.cards)
        return cards_dealt >= self.cut_card_position

    def draw(self) -> int:
        """Draw a card from the shoe."""
        if not self.cards:
            self.shuffle()
        return self.cards.pop()

    def draw_specific(self, value: int) -> Optional[int]:
        """
        Draw a specific card value from the shoe (for testing/setup).
        Returns None if card not available.
        """
        if value in self.cards:
            self.cards.remove(value)
            return value
        return None


class InfiniteDeck:
    """
    Infinite deck - each card drawn independently with equal probability.
    Used for computing base strategy (no card counting effects).
    """

    # Probability weights for each card value
    # 2-9: 1/13 each, 10/J/Q/K: 4/13, A: 1/13
    WEIGHTS = np.array([1, 1, 1, 1, 1, 1, 1, 1, 4, 1], dtype=np.float64)
    WEIGHTS = WEIGHTS / WEIGHTS.sum()
    VALUES = np.array([2, 3, 4, 5, 6, 7, 8, 9, 10, 11])

    def __init__(self):
        """Initialize infinite deck."""
        pass

    def shuffle(self) -> None:
        """No-op for infinite deck."""
        pass

    def needs_shuffle(self) -> bool:
        """Infinite deck never needs shuffle."""
        return False

    def draw(self) -> int:
        """Draw a random card with correct probability distribution."""
        return int(np.random.choice(self.VALUES, p=self.WEIGHTS))

    def draw_specific(self, value: int) -> int:
        """For infinite deck, always returns the requested value."""
        return value


def hand_value(cards: List[int]) -> tuple:
    """
    Calculate the value of a hand.

    Args:
        cards: List of card values (Ace = 11)

    Returns:
        (total, is_soft): Total value and whether hand is soft
    """
    total = sum(cards)
    aces = cards.count(11)

    # Convert aces from 11 to 1 as needed to avoid bust
    while total > 21 and aces > 0:
        total -= 10
        aces -= 1

    # Hand is soft if it contains an ace counted as 11
    is_soft = aces > 0 and total <= 21

    return total, is_soft


def is_blackjack(cards: List[int]) -> bool:
    """Check if hand is a natural blackjack (Ace + 10-value on first 2 cards)."""
    if len(cards) != 2:
        return False
    total, _ = hand_value(cards)
    return total == 21


def is_bust(cards: List[int]) -> bool:
    """Check if hand is busted (over 21)."""
    total, _ = hand_value(cards)
    return total > 21


def is_pair(cards: List[int]) -> bool:
    """Check if hand is a splittable pair."""
    return len(cards) == 2 and cards[0] == cards[1]


def get_state(player_cards: List[int], dealer_upcard: int) -> tuple:
    """
    Get the state tuple for strategy lookup.

    Args:
        player_cards: Player's cards
        dealer_upcard: Dealer's visible card

    Returns:
        (player_total, dealer_upcard, is_soft, is_pair)
    """
    total, is_soft = hand_value(player_cards)
    pair = is_pair(player_cards)

    return (total, dealer_upcard, is_soft, pair)

File: rust/src/deck.rs
//! Deck/Shoe management for Blackjack simulation.
//! Optimized for speed with fixed-size arrays and fast RNG.

use fastrand::Rng;

/// Maximum cards in a hand (5 cards + safety margin)
pub const MAX_HAND_SIZE: usize = 12;

/// Fixed-size hand to avoid heap allocations
#[derive(Clone, Copy)]
pub struct Hand {
    cards: [u8; MAX_HAND_SIZE],
    len: u8,
}

impl Hand {
    #[inline(always)]
    pub fn new() -> Self {
        Hand {
            cards: [0; MAX_HAND_SIZE],
            len: 0,
        }
    }

    #[inline(always)]
    pub fn from_cards(c1: u8, c2: u8) -> Self {
        let mut h = Hand::new();
        h.cards[0] = c1;
        h.cards[1] = c2;
        h.len = 2;
        h
    }

    #[inline(always)]
    pub fn push(&mut self, card: u8) {
        self.cards[self.len as usize] = card;
        self.len += 1;
    }

    #[inline(always)]
    pub fn len(&self) -> usize {
        self.len as usize
    }

    #[inline(always)]
    pub fn cards(&self) -> &[u8] {
        &self.cards[..self.len as usize]
    }

    #[inline(always)]
    pub fn first(&self) -> u8 {
        self.cards[0]
    }

    #[inline(always)]
    pub fn second(&self) -> u8 {
        self.cards[1]
    }
}

impl Default for Hand {
    fn default() -> Self {
        Self::new()
    }
}

/// Infinite deck with fast RNG
/// Uses lookup table for O(1) card drawing
pub struct InfiniteDeck {
    rng: Rng,
}

// Lookup table: maps random value 0-12 to card value
// 0-7 -> 2-9, 8-11 -> 10, 12 -> 11 (Ace)
const CARD_LOOKUP: [u8; 13] = [2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 11];

impl InfiniteDeck {
    #[inline(always)]
    pub fn new() -> Self {
        InfiniteDeck {
            rng: Rng::new(),
        }
    }

    /// Draw a random card - O(1) with lookup table
    #[inline(always)]
    pub fn draw(&mut self) -> u8 {
        CARD_LOOKUP[self.rng.usize(0..13)]
    }
}

impl Default for InfiniteDeck {
    fn default() -> Self {
        Self::new()
    }
}

/// Calculate hand value - optimized with early exit
#[inline(always)]
pub fn hand_value(hand: &Hand) -> (u8, bool) {
    let cards = hand.cards();
    let mut total: u16 = 0;
    let mut aces: u8 = 0;

    for &card in cards {
        total += card as u16;
        aces += (card == 11) as u8;
    }

    // Convert aces from 11 to 1 as needed
    while total > 21 && aces > 0 {
        total -= 10;
        aces -= 1;
    }

    (total as u8, aces > 0)
}

/// Check if hand is a natural blackjack
#[inline(always)]
pub fn is_blackjack(hand: &Hand) -> bool {
    hand.len() == 2 && {
        let (total, _) = hand_value(hand);
        total == 21
    }
}

/// Check if hand is busted
#[inline(always)]
pub fn is_bust(hand: &Hand) -> bool {
    let (total, _) = hand_value(hand);
    total > 21
}

/// Player state for strategy lookup
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct PlayerState {
    pub total: u8,
    pub dealer_upcard: u8,
    pub is_soft: bool,
    pub is_pair: bool,
}

impl PlayerState {
    #[inline(always)]
    pub fn new(total: u8, dealer_upcard: u8, is_soft: bool, is_pair: bool) -> Self {
        PlayerState { total, dealer_upcard, is_soft, is_pair }
    }
}

/// Generate starting hand for a state
#[inline(always)]
pub fn get_hand_for_state(total: u8, is_soft: bool, is_pair: bool) -> Hand {
    if is_pair {
        if is_soft {
            Hand::from_cards(11, 11) // A,A
        } else {
            let card = total / 2;
            Hand::from_cards(card, card)
        }
    } else if is_soft {
        Hand::from_cards(11, total - 11)
    } else if total <= 11 {
        if total >= 4 {
            Hand::from_cards(2, total - 2)
        } else {
            Hand::from_cards(total, 0) // edge case
        }
    } else if total <= 19 {
        Hand::from_cards(10, total - 10)
    } else {
        Hand::from_cards(10, 10) // 20
    }
}
